runtime/caml/platform.h

```diff
@@ -61,6 +61,11 @@ Caml_inline uintnat atomic_load_relaxed(atomic_uintnat* p)
   return atomic_load_explicit(p, memory_order_relaxed);
 }
 
+Caml_inline value atomic_load_relaxed_val(value* p)
+{
+  return atomic_load_explicit(Op_atomic_val(p), memory_order_relaxed);
+}
```

Comment:
Instead of duplicating the function to work with different signatures, would it help to make it a macro so that it is polymorphic? (e.g. as in https://github.com/gadmm/ocaml/compare/custom_thread_safety2..custom_thread_safety)
Would that help with writing the code differently?

Reply:
If I understand correctly, the race with `*me.start` is because the value resides on the OCaml heap, and thus can be mutated in parallel by OCaml code. 1) Good catch, 2) In this case, I think the way it is done elsewhere is that the type of pointers to values that reside in the OCaml heap is never `value *`, but `volatile value *` or even `volatile const value *`.

Assume we change the types of `mark_entry` elements to `volatile const value *`. This fixes the race. Is this still a problem for OCaml-TSan, or for using TSan to debug the runtime? Because I suspect this is far from the only place where the issue will arise.

- If OCaml-TSan: can we simply assume the GC is correct and exclude this function from TSan instrumentation?
- If TSan used to debug the runtime: is it not a bit hopeless to try to get rid of reported races on volatile, given how much code is written with volatile?

Reply:
E.g. if we make it atomic here, we would like to apply this style consistently in the runtime; at the same time we cannot change the `Field` macro. Do you see a clear boundary between what should be made atomic, and what must remain volatile, based on TSan needs? For instance, we could say that only the C code which is part of the mutator (parts of the C runtime called from OCaml) should be allowed to use volatile. Would it improve the situation?

Reply:
My take:
1. The TSan implmentation in either GCC and Clang is experimental, poorly documented and difficult to use; it works well for simple C programs following the C11 memory model to the letter, and poorly outside this subset. The implementation is full of odd special cases for other languages than C.
2. While the pure-C parts of the C runtime use the C11 memory model, the FFI parts of the runtime (the parts that interact with the OCaml heap) are a volatile-based model that @gadmm suggested and corresponds to some parts of the Linux Kernel approach.
3. TSan does not currently support volatile models well enough.
4. The solution is to improve TSan -- which is a service not just to OCaml but to a much wider community of programmers, including the Linux kernel.

My understanding is that @OlivierNicole has a slightly different view of things (maybe motivated by the idea of using the tools as they are and not broadening the scope of his work too much in the direction of "fix both GCC and Clang"), where he thinks of the `volatile` use in the OCaml runtime as a "deprecated" pattern -- despite the fact that there we offer no support for anything else than this approach right now -- so we could look the other way and hope to eventually convert the runtime to be fully C11-memory-model based, including the cross-language parts.